Fix ident in XML representation of VMTemplate

Ulme007 opened the pull request:
Hi, 
I'm using Your great OpenNebula RPC implementation, but I've found an error in the XML representation of the VMTemplate. An iteration of all templates shows that the ident of the VMTemplate is allways '0'. I changed the ident to 'ID'. Now it works fine.
Cheers Ulli.

Commit:
Fix ident in XML representation of VMTemplate

marthjod commented:
Hi @Ulme007, thanks for your PR. I've been meaning to merge the cleaner [rm-ocatypes](https://github.com/marthjod/gocart/tree/rm-ocatypes) branch into master, this is where the new code lives. Can you reproduce your error using this branch as well and if so, could you please post the PR against the feature branch, I'd be happy to merge it then. Also, could you check whether the ID is always `ID`? I've stumbled upon cases where ID keys were different depending on where they occur (cf. [here](https://github.com/marthjod/gocart/commit/b5645a05d904e19b74f9da79b62fc3e693af9937)).

Ulme007 commented:
Hi,
I changed the dep configuration file to get your implementation from 'rm-ocatypes' branch and I also changed my implementation to get information about 'VMs', 'Templates' and 'Images'. Everything works fine. The problem with the template ID is fixed in this branch. Danke schön!
What are the next steps? Do you plan to merge this branch into master? If so, we need to reimplement our tools. 
And by the way. Can you add a license file? This helps a lot, that I can use the library without bad conscience.

marthjod commented:
I've rebased `rm-ocatypes` onto `master` and will obsolete it shortly. You don't need to reimplement anything, you can pin an older ref in your dependency management if you have to. Nevertheless I'd recommend a switch which is very easy since the API hasn't changed significantly (refer to the examples for how to issue calls now, the rest is basically the same). I've added a licence file.

Ulme007 commented:
I switch to german. To describe everything in english takes too much time.

Danke, dass Du ein Lizenzfile hinzugefügt hast!

Du hast recht, wir können mit unserer Implementation auf dem gepinnten Master bleiben. Ich bin gerade dabei eine anwendung zu schreiben, die mir ungenutzte, d.h. unreferenzierte Images und Templates ausgibt, um diese dann später zu löschen. Jetzt habe ich sowohl im master als auch im rm-ocatypes Branch das Problem, dass von den Templates nicht die Disks geliefert bekomme.

Ich mache


vt := vmtplpool.NewVMTemplatePool()
for _, template := range vt.Templates {
    fmt.Printf("Disk len: %d\n", len(template.Disk))
}


und bekomme bei allen Templates eine 'Disk len' von 0. Eigentlich würde ich erwarten, das hier die referenzierten Images ausgegeben werden. Hast Du eine Idee?

marthjod commented:
Wie sieht denn der (bereinigte/gekürzte) Input aus? `onetemplate list -x` o.ä.

Ulme007 commented:
`<VMTEMPLATE_POOL>
  <VMTEMPLATE>
    <ID>271</ID>

    <TEMPLATE>

      <DISK>
        <IMAGE><![CDATA[centos7-60g_2018-10-10-08-19-49+0000_2ee14b36d03e3b61cf9d3403042aaa1b87b3cac5]]></IMAGE>
        <IMAGE_UNAME><![CDATA[oneadmin]]></IMAGE_UNAME>
      </DISK>
      
    </TEMPLATE>
  </VMTEMPLATE>

</VMTEMPLATE_POOL>`

marthjod commented:
Ok, I'll have a look later. Feel free to open an issue for this.

Ulme007 commented:
Die disk in einem Template kann auch unterschiedlich angegeben werden:
`
<DISK>
   <DEV_PREFIX><![CDATA[vd]]></DEV_PREFIX>
   <IMAGE_ID><![CDATA[119]]></IMAGE_ID>
</DISK>
<DISK>
    <DEV_PREFIX><![CDATA[vd]]></DEV_PREFIX>
    <DRIVER><![CDATA[qcow2]]></DRIVER>
    <IMAGE><![CDATA[review.server-dev.de_data]]></IMAGE>
    <IMAGE_UNAME><![CDATA[oneadmin]]></IMAGE_UNAME>
</DISK>
`
Einmal als IMAGE_ID und einmal nur als IMAGE.

Ulme007 commented:
Ich habe den Fehler gefunden. Er war auf meiner Seite!

	for _, template := range templatePool.Templates {
		fmt.Printf("Disk len: %d\n", len(template.Template.Disk))
	}

Das 'template' aus dem TemplatePool muss ich noch weiter mit dem HostTemplate referenzieren, dann klappt es. Also 'template.Template.Disk' anstatt 'template.Disk'.

Danke Dir für Deine Hilfe!
